### DISCUSSION TITLE
Good practices for security: limited capabilities on a task?

### DISCUSSION THREAD

**Discussion by rosenhouse** [concourse/concourse#7342] [Help & Support] [answered]
I have a soon-to-be open-source project, and am setting up Concourse CI to provide status checks on PRs on GitHub.

The test binary needs to run with [the `IPC_LOCK` capability](https://man7.org/linux/man-pages/man7/capabilities.7.html).  As best I can tell, to grant the binary this capability, the Concourse task must run with `privileged: true`.  

**Q**: Do I have that right?

I see https://github.com/concourse/concourse/discussions/6912 proposes an interface for limiting capabilities of Concourse containers, but I don't see anything like this that is usable today.

I'm thinking through the security of the CI system.  I believe that if I run the Concourse task with `privileged: true`, then PRs to the project could pose a security risk (e.g. by adding code to break out of the test container).

As a workaround, I'm considering the following:
1. The pipeline is managed outside the open-source repo (the repo may contain a copy of the pipeline config for visibility, but that isn't the source of truth, so PRs to that config file cannot change the execution environment of the tests)
2. The pipeline runs an inlined task, with `privileged: true` which runs a inlined shell script to
    - immediately drop all capabilities except `CAP_IPC_LOCK`
    - then launches a `test.sh` script from the repository.

**Q**: Is that a reasonable approach?  Are there additional risks to be aware of here?

**Q**: Is there any prior-art for this sort of thing?  I'd love to follow an established pattern if there is one.

Thank you!

**Comment by taylorsilva** [comment, marked as answer]
You are correct, currently there is no way to grant individual capabilities to task containers. It's all or nothing with `privileged`.

Your workaround makes sense to me. Not aware of any prior art that I can point you to.
